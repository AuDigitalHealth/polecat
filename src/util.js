// SHA-256 function from
// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest
export async function sha256(str) {
  // We transform the string into an arraybuffer.
  const buffer = str2ab(str)
  return crypto.subtle.digest('SHA-256', buffer).then(function(hash) {
    return hex(hash)
  })
}

function str2ab(str) {
  const buf = new ArrayBuffer(str.length * 2) // 2 bytes for each char
  const bufView = new Uint16Array(buf)
  for (let i = 0, strLen = str.length; i < strLen; i++) {
    bufView[i] = str.charCodeAt(i)
  }
  return buf
}

function hex(buffer) {
  const hexCodes = []
  const view = new DataView(buffer)
  for (let i = 0; i < view.byteLength; i += 4) {
    // Using getUint32 reduces the number of iterations needed (we process 4
    // bytes each time)
    const value = view.getUint32(i)
    // toString(16) will give the hex representation of the number without
    // padding
    const stringValue = value.toString(16)
    // We use concatenation and slice for padding
    const padding = '00000000'
    const paddedValue = (padding + stringValue).slice(-padding.length)
    hexCodes.push(paddedValue)
  }

  // Join all the hex strings into one
  return hexCodes.join('')
}

export const formatNumber = number => {
  if (typeof number !== 'number') return undefined
  const numString = number.toString()
  return [...numString].reduce((prev, curr, i, arr) => {
    return arr.length - i - 1 !== 0 && (arr.length - i - 1) % 3 === 0
      ? prev + curr + ','
      : prev + curr
  }, '')
}

export const parseUri = str => {
  var o = parseUri.options,
    m = o.parser[o.strictMode ? 'strict' : 'loose'].exec(str),
    uri = {},
    i = 14

  while (i--) uri[o.key[i]] = m[i] || ''

  uri[o.q.name] = {}
  uri[o.key[12]].replace(o.q.parser, function($0, $1, $2) {
    if ($1) uri[o.q.name][$1] = $2
  })

  return uri
}

parseUri.options = {
  strictMode: false,
  key: [
    'source',
    'protocol',
    'authority',
    'userInfo',
    'user',
    'password',
    'host',
    'port',
    'relative',
    'path',
    'directory',
    'file',
    'query',
    'anchor',
  ],
  q: {
    name: 'queryKey',
    parser: /(?:^|&)([^&=]*)=?([^&]*)/g,
  },
  parser: {
    // eslint-disable-next-line no-useless-escape
    strict: /^(?:([^:\/?#]+):)?(?:\/\/((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?))?((((?:[^?#\/]*\/)*)([^?#]*))(?:\?([^#]*))?(?:#(.*))?)/,
    // eslint-disable-next-line no-useless-escape
    loose: /^(?:(?![^:@]+:[^:@\/]*@)([^:\/?#.]+):)?(?:\/\/)?((?:(([^:@]*)(?::([^:@]*))?)?@)?([^:\/?#]*)(?::(\d*))?)(((\/(?:[^?#](?![^?#\/]*\.[^?#\/.]+(?:[?#]|$)))*\/?)?([^?#\/]*))(?:\?([^#]*))?(?:#(.*))?)/,
  },
}
