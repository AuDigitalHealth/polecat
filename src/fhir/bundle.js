import {
  getSubjectConcept,
  codingToSnomedCode,
  relationshipTypeFor,
  amtConceptTypeFor,
  groupUri,
  emptyConcepts,
} from './medication.js'
import { queryFromSearchObject } from './search.js'
import { sha256 } from '../util.js'

// Get concepts from the supplied bundle, relating them to the subject concept.
export async function getBundleConcepts(subject, bundle, options = {}) {
  const { groupingThreshold = 3, queryType = 'children' } = options
  if (!bundle || bundle.total === 0) return emptyConcepts()
  if (bundle.total <= groupingThreshold) {
    return bundle.entry.reduce((acc, e) => {
      const child = getSubjectConcept(e.resource)
      acc.concepts.push(child)
      acc.relationships.push({
        source: codingToSnomedCode(child.coding),
        target: codingToSnomedCode(subject.coding),
        type: relationshipTypeFor(child.type, subject.type),
      })
      return acc
    }, emptyConcepts())
  } else {
    const concepts = bundle.entry
      .slice(0, groupingThreshold)
      .map(e => getSubjectConcept(e.resource))
    // The group's code is a hash of the concept data within the group.
    const groupCode = await sha256(JSON.stringify(concepts)),
      // Group relationship type defaults to the relationship type inferred from
      // the first concept in the group as the source, and the subject concept
      // as the target.
      {
        groupRelationshipType = relationshipTypeFor(
          concepts[0].type,
          subject.type,
        ),
      } = options,
      query = queryForGrouping(subject.coding, queryType, concepts[0].type)
    return {
      concepts: [
        {
          coding: [{ system: groupUri, code: groupCode }],
          type: 'group',
          total: bundle.total,
          query,
          concepts,
        },
      ],
      relationships: [
        {
          // A `group-` prefix is added to the group code within the
          // relationships so that it can be discerned when drawing curves and
          // arrows.
          source: `group-${groupCode}`,
          target: codingToSnomedCode(subject.coding),
          type: groupRelationshipType,
          // This flag is set so that we know to pluralise link labels.
          sourceIsGroup: true,
        },
      ],
    }
  }
}

const queryForGrouping = (coding, queryType, childType) => {
  switch (queryType) {
    case 'packages':
      return queryFromSearchObject({
        package: coding,
        type: amtConceptTypeFor(childType),
      })
    case 'children':
      return queryFromSearchObject({
        parent: coding,
        type: amtConceptTypeFor(childType),
      })
    default:
      throw new Error('Unknown query type')
  }
}

export const nextLinkFromBundle = bundle => {
  if (!bundle || !bundle.link) return null
  const next = bundle.link.find(l => l.relation === 'next')
  if (!next || !next.url) return null
  return next.url
}

export const previousLinkFromBundle = bundle => {
  if (!bundle || !bundle.link) return null
  const previous = bundle.link.find(l => l.relation === 'previous')
  if (!previous || !previous.url) return null
  return previous.url
}
