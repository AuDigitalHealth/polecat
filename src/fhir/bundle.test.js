import sinon from 'sinon'

import { getBundleConcepts } from './bundle.js'
import * as util from '../util.js'

import bundle from '../../test/childBundle-21360011000036101.json'

describe('getBundleConcepts', () => {
  beforeEach(() => {
    // Need to stub out `sha256` function, as it relies on APIs only available
    // within the browser.
    sinon.stub(util, 'sha256').callsFake(() => Promise.resolve('somesha'))
  })

  it('should return correct concepts', () => {
    const subject = {
      coding: [
        {
          system: 'http://snomed.info/sct',
          code: '21360011000036101',
          display: 'amoxicillin + clavulanic acid',
        },
      ],
      type: 'UPD',
    }
    const result = getBundleConcepts(subject, bundle)
    expect(result).resolves.toMatchSnapshot()
  })

  afterEach(() => {
    util.sha256.restore()
  })
})
