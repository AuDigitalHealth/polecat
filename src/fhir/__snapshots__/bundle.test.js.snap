// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBundleConcepts should return correct concepts 1`] = `
Object {
  "concepts": Array [
    Object {
      "coding": Array [
        Object {
          "code": "somesha",
          "system": "group",
        },
      ],
      "concepts": Array [
        Object {
          "coding": Array [
            Object {
              "code": "975591000168101",
              "display": "amoxicillin 1 g + clavulanic acid 200 mg injection, 1.2 g vial",
              "system": "http://snomed.info/sct",
            },
          ],
          "type": "UPDSF",
        },
        Object {
          "coding": Array [
            Object {
              "code": "975601000168108",
              "display": "Amoxiclav 1000/200 (Juno) powder for injection, 1.2 g vial",
              "system": "http://snomed.info/sct",
            },
          ],
          "type": "BPSF",
        },
        Object {
          "coding": Array [
            Object {
              "code": "23463011000036105",
              "display": "amoxicillin 400 mg/5 mL + clavulanic acid 57 mg/5 mL powder for oral liquid",
              "system": "http://snomed.info/sct",
            },
          ],
          "type": "UPDSF",
        },
      ],
      "query": "type:MPUU parent:\\"21360011000036101|amoxicillin + clavulanic acid\\"",
      "total": 79,
      "type": "group",
    },
  ],
  "relationships": Array [
    Object {
      "source": "group-somesha",
      "sourceIsGroup": true,
      "target": "21360011000036101",
      "type": "is-a",
    },
  ],
}
`;
